File: PalindromicRanges/py2/pr.py
# palindromic ranges soln in py2 for code eval by steven a dunn

import sys

def is_palindrome(x):
	x_str = str(x)
	if x_str == x_str[::-1]:
		return True
	return False 

def get_num_ranges(x, y):
	interesting_range_count = 0
	nums = []
	for i in range(x, y+1):
		if is_palindrome(i):
			nums.append(1)
		else:
			nums.append(0)
	for i in range(len(nums)):
		palindrome_count = 0
		for j in range(i, len(nums)):
			if nums[j] == 1:
				palindrome_count += 1
			if palindrome_count % 2 == 0:
				interesting_range_count += 1
	return interesting_range_count

f = open(sys.argv[1], 'r')
for line in f:
	x, y = map(int, line.strip().split())
	print get_num_ranges(x, y)
f.close()


"""

bool isPalindrome(int x)
{
  string word;
  ostringstream oss;
  oss << x;
  word = oss.str();
  
  string reverseWord = word;
  reverse(reverseWord.begin(), reverseWord.end());

  if (word == reverseWord)
    return true;

  return false;
}
"""

File: PalindromicRanges/cpp/PR.cpp
// Palindromic Range solution in C++ for CodeEval.com by Steven A. Dunn

#include <algorithm>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using std::cout;
using std::endl;
using std::ifstream;
using std::ostringstream;
using std::reverse;
using std::string;
using std::vector;

int numRanges(string);
bool isPalindrome(int);

int main(int argc, char* argv[])
{
  ifstream inputFile(argv[1]);
  string line;

  if(inputFile)
  {
    while(getline(inputFile, line))
    {
      int rangeCount = numRanges(line);
      cout << rangeCount << endl;
    }
    inputFile.close();
  }

  return 0;
}

int numRanges(string rangeStr)
{
  int x = atoi(rangeStr.substr(0, rangeStr.find(" ")).c_str());
  int y = atoi(rangeStr.substr(rangeStr.find(" ")).c_str());

  int interestingRangeCount = 0;
 
  vector<int> nums;
  for(int i = x; i <= y; ++i)
  {
    if(isPalindrome(i))
      nums.push_back(1);
    else
      nums.push_back(0);
  }

  for(int i = 0; i < nums.size(); ++i)
  {
    int palindromeCount = 0;
    for(int j = i; j < nums.size(); ++j)
    {
      if(nums[j] == 1)
         ++palindromeCount;

      if(palindromeCount % 2 == 0)
        ++interestingRangeCount;
    }
  }

  return interestingRangeCount;
}

bool isPalindrome(int x)
{
  string word;
  ostringstream oss;
  oss << x;
  word = oss.str();
  
  string reverseWord = word;
  reverse(reverseWord.begin(), reverseWord.end());

  if (word == reverseWord)
    return true;

  return false;
}
